### PR TITLE
Add emojis (custom and native)

### PR DESCRIPTION
**Pull request by jonnhy12333**
Okay so this is finally ready for pull request (but probs not).  Here are some summary items about what this idiot pull request is (in no particular order):
 - Added emoji picker / icon to the input allowing user to insert emojis at the current cursor position
 - Added native / custom emoji rendering support to the user mood and messages
 - Added the ability to add custom emojis via the app menu (restrictions built-in to not allow bad names / names that already exist as a custom or native emoji)
 - No resizing of the emoticon will be done (couldn't find a good way to do this to gifs)
 - In the future we could allow deleting your own uploaded custom emojis (added the 'uid' field so that we know who uploaded the emoticon as well as an 'added' date so that you could see your most recent uploaded emojis)
 - No sort of custom emojis is being done currently (the order you see them in the picker is the order they were uploaded)
 - New pipe added to transform text input using our custom emojis.  Note that to get the full effect, we use the 'emojify' pipe (from angular/emojify for native emojis) and our new 'customEmojify' pipe for our emojis
 - Emojis are uploaded into the firebase database, so different dbs have different emojis
 - Emojis that are sent as single (just 1 emoji) will display as large (128x128), otherwise 32x32
 - No auto-complete has been done for the text input, shouldn't be too hard in the future though

THE QUESTION IS:
 - Assuming this passes, who should we let add custom emojis?  I can see *some* people abusing this and I don't think everyone should have the right to do it at first.....

### CODE REVIEW

**Comment by jonnhy12333** [comment]
Here is how you access the custom emojis:
image

And obviously here is how you add custom emoji:
image

This dialog will let you enter the unique name, description and select the image (showing both large and small previews).  Ignore the error for now, there is already a 'hylanld' emoji...
image

File: src/app/directives/short-name-validator.directive.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+  validate(control: AbstractControl): ValidationErrors {
+
+    if (!control.value) {
+      return {
+        shortName: 'Name is required',
+      };
+    }
+
+    if (control.value.split(' ').length > 1) {
```
Comment (greggiannone) [inline src/app/directives/short-name-validator.directive.ts]: You should probably have this validate it doesn't match a regex with the whitespace matcher to catch dickery around tabs and maybe newlines

File: src/app/directives/short-name-validator.directive.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  doesDefaultEmojiExist(shortName: string): boolean {
+    for (const e of this.emojiSearch.search(shortName)) {
+      for (const s of e.shortNames) {
+        if (shortName === s) {
+          return true;
+        }
+      }
+    }
```
Comment (greggiannone) [inline src/app/directives/short-name-validator.directive.ts]: This method doesn't actually return if it gets out of the loop

File: src/app/directives/short-name-validator.directive.ts (hunk 1 above)
Comment (jonnhy12333) [reply src/app/directives/short-name-validator.directive.ts]: yep, using regex now

File: src/app/components/add-emoticon-dialog/add-emoticon-dialog.component.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        this.dialogRef.close();
+      })
+      .finally(() => {
+        this.isSaving = false;
+      });
+    }
+  }
+
+  onEmojiChanged(file: File): void {
+    console.log('file is', file);
```
Comment (greggiannone) [inline src/app/components/add-emoticon-dialog/add-emoticon-dialog.component.ts]: Left a log in here

File: src/app/directives/short-name-validator.directive.ts (hunk 1 above)
Comment (jonnhy12333) [reply src/app/directives/short-name-validator.directive.ts]: as well as the regex you suggested to ensure there are only characters and numbers / etc. from the real regex

File: src/app/directives/short-name-validator.directive.ts (hunk 2 above)
Comment (jonnhy12333) [reply src/app/directives/short-name-validator.directive.ts]: Whoops, fixed

File: src/app/pipes/custom-emojify.pipe.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+  transform(value: string, args?: any): any {
+    let newValue = value;
+    let matches = this.emoticonPattern.exec(newValue);
+    while (matches) {
+      // Show large emojis when the message is only 1 emoji
+      const isOnlyEmoji = value.length === matches[0].length;
+      const emoji = this.chat.customEmoticonDictionary[matches[1]];
+      if (emoji) {
+        let cssClass = 'hmct-message__emoticon--small';
```
[the remark this answers is not in the document]
Comment (jonnhy12333) [reply src/app/pipes/custom-emojify.pipe.ts]: Looked into it, wasn't easy, leaving at default font size for now, maybe future enhancement

File: src/app/components/add-emoticon-dialog/add-emoticon-dialog.component.ts (hunk 3 above)
Comment (jonnhy12333) [reply src/app/components/add-emoticon-dialog/add-emoticon-dialog.component.ts]: Whoops, removed

File: src/app/components/add-emoticon-dialog/add-emoticon-dialog.component.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    theme: ThemeService,
+    private auth: AuthService,
+    private fbStorage: AngularFireStorage,
+  ) { }
+
+  ngOnInit() {
+  }
+
+  uploadEmoji(): void {
+    if (this.addEmojiForm.valid) {
```
[the remark this answers is not in the document]
Comment (jonnhy12333) [reply src/app/components/add-emoticon-dialog/add-emoticon-dialog.component.ts]: Yeah this is an odd issue, sometimes you have to search for weird terms for it to show up.  Seems after a refresh that sometimes fixes the search, but the search results don't seem to be dynamic and there is no available method to refresh it.  Seems this was built with the idea that custom emojis would be set at the start and not change.  The good news is that at least the picker has em and users should be able to see the new emojis when sent.

File: src/app/components/add-emoticon-dialog/add-emoticon-dialog.component.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        this.dialogRef.close();
+      })
+      .finally(() => {
+        this.isSaving = false;
+      });
+    }
+  }
+
+  onEmojiChanged(file: File): void {
+    const fileReader = new FileReader();
```
[the remark this answers is not in the document]
Comment (jonnhy12333) [reply src/app/components/add-emoticon-dialog/add-emoticon-dialog.component.ts]: Yep, we already talked about the deets, but we're now doing that

**Pull request merged**